algorithm/cgal/Loft2.h

```diff
@@ -0,0 +1,321 @@
+        negative_layers.push_back(std::move(negative_cycles));
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+  }
+  if (positive_layers.size() < 2) {
+    std::cout << "Loft: needs at least two positive layers." << std::endl;
+    return STATUS_EMPTY;
+  }
+
+  std::vector<EK::Point_3> positive_points;
+  std::vector<std::vector<size_t>> positive_faces;
+  for (size_t nth = 0; nth + 1 < positive_layers.size(); nth++) {
+    loft_between_layers(positive_layers[nth], positive_layers[nth + 1],
+                        positive_points, positive_faces);
+  }
+
+  std::vector<EK::Point_3> negative_points;
+  std::vector<std::vector<size_t>> negative_faces;
+  for (size_t nth = 0; nth + 1 < negative_layers.size(); nth++) {
+    loft_between_layers(negative_layers[nth], negative_layers[nth + 1],
+                        negative_points, negative_faces);
+  }
+
+  auto build_mesh = [&](std::vector<EK::Point_3>& points,
+                        std::vector<std::vector<size_t>>& faces, bool close,
+                        Surface_mesh& mesh) {
+    CGAL::Polygon_mesh_processing::repair_polygon_soup(points, faces);
+
+    // Unfortunately this may be inverted.
+    // If we close the mesh it will be re-oriented to bound the volume.
+    // Otherwise the orientation is accidental.
+    // FIX: Could we intersect the surface with a ray to figure out the winding
+    // order?
+    CGAL::Polygon_mesh_processing::orient_polygon_soup(points, faces);
+
+    CGAL::Polygon_mesh_processing::polygon_soup_to_polygon_mesh(points, faces,
+                                                                mesh);
+
+    if (close) {
+      std::vector<typename Surface_mesh::halfedge_index> border_cycles;
+      CGAL::Polygon_mesh_processing::extract_boundary_cycles(
+          mesh, std::back_inserter(border_cycles));
+      for (const Surface_mesh::Halfedge_index edge : border_cycles) {
+        std::vector<Surface_mesh::Face_index> faces;
+        CGAL::Polygon_mesh_processing::triangulate_hole(
+            mesh, edge, std::back_inserter(faces),
+            CGAL::parameters::use_2d_constrained_delaunay_triangulation(false));
+      }
+      if (CGAL::is_closed(mesh)) {
+        CGAL::Polygon_mesh_processing::orient_to_bound_a_volume(mesh);
+      }
+    }
+  };
+
+  Surface_mesh positive_mesh;
+  build_mesh(positive_points, positive_faces, close, positive_mesh);
+
+  Surface_mesh negative_mesh;
+  build_mesh(negative_points, negative_faces, close, negative_mesh);
+
+  if (CGAL::is_closed(positive_mesh)) {
+    if (!cut_mesh_by_mesh(positive_mesh, negative_mesh)) {
+      return STATUS_ZERO_THICKNESS;
+    }
+  } else {
+    CGAL::Polygon_mesh_processing::reverse_face_orientations(positive_mesh);
+    positive_mesh.join(negative_mesh);
+  }
+
+  int target = geometry->add(GEOMETRY_MESH);
+  geometry->setIdentityTransform(target);
+  demesh(positive_mesh);
+  geometry->mesh(target) = std::move(positive_mesh);
+  return STATUS_OK;
+}
```

Comment:
This is a super cool loft algorithm. Did you find it somewhere or create it?

Reply:
Unfortunately I had to figure it out, which is why it took a while :)

The basic idea is pretty easy -- you find the best triangle to bridge the gap between the loops.
The hard part is making sure that it doesn't intersect with any existing triangles.

What I eventually realized is that you can think of each triangle formation as a cut operation.
The first cut splits the loops into strips.
After that, each new triangle splits that strip into two sub-strips.

This also allows the best triangles to be found independently, which avoids build-up error.

But there are still some vulnerabilities.

Any number of triangles can meet at the same tip, so if you have inclined surfaces being lofted, all of the bottom surface will want to join at the bottom-most point of the top surface, and so on.

Having them take turns seems to help break up these patterns, but there are still degenerate cases coming up.

Reply:
That is super cool! You should publish that idea somewhere because I think that has some general applicability

Reply:
Might be an idea, but should understand these failure cases better first. :)